sycl/include/sycl/detail/type_list.hpp

```diff
@@ -12,53 +12,26 @@
 
 #include <type_traits> // for bool_constant, conditional_t, fals...
 
+#include <sycl/detail/boost/mp11/algorithm.hpp>
+#include <sycl/detail/boost/mp11/set.hpp>
+
 namespace sycl {
 inline namespace _V1 {
 namespace detail {
 
-template <typename T> using head_t = typename T::head;
-
-template <typename T> using tail_t = typename T::tail;
-
-// type_list
-template <typename... T> struct type_list;
+template <class... T> using type_list = boost::mp11::mp_list<T...>;
 
 using empty_type_list = type_list<>;
 
 template <typename T>
-struct is_empty_type_list
-    : std::conditional_t<std::is_same_v<T, empty_type_list>, std::true_type,
-                         std::false_type> {};
-
-template <> struct type_list<> {};
-
-template <typename Head, typename... Tail> struct type_list<Head, Tail...> {
-  using head = Head;
-  using tail = type_list<Tail...>;
-};
-
-template <typename Head, typename... Tail, typename... Tail2>
-struct type_list<type_list<Head, Tail...>, Tail2...> {
-private:
-  using remainder = tail_t<type_list<Head>>;
-  static constexpr bool has_remainder = !is_empty_type_list<remainder>::value;
-  using without_remainder = type_list<Tail..., Tail2...>;
-  using with_remainder = type_list<remainder, Tail..., Tail2...>;
-
-public:
-  using head = head_t<type_list<Head>>;
-  using tail =
-      std::conditional_t<has_remainder, with_remainder, without_remainder>;
-};
+using is_empty_type_list = std::is_same<T, empty_type_list>;
 
 // is_contained
 template <typename T, typename TypeList>
-struct is_contained
-    : std::conditional_t<std::is_same_v<std::remove_cv_t<T>, head_t<TypeList>>,
-                         std::true_type, is_contained<T, tail_t<TypeList>>> {};
+using is_contained =
+    boost::mp11::mp_set_contains<TypeList, std::remove_cv_t<T>>;
 
-template <typename T>
-struct is_contained<T, empty_type_list> : std::false_type {};
+using boost::mp11::mp_append;
```

Comment:
what name do you suggest? Our current naming (e.g. `is_contained`) doesn't contain any info about it being associated with `type_list`. If we follow that naming it would be called `append`. But we might want some other `append` which isn't for `type_list` in `detail` in the future. That's why I thought having a prefix might be better. Do you prefer a different prefix? E.g. tl (for type_list) and name it `tl_append`.

Reply:
I am good with `tl_append` or `type_list_append` (latter being a little more verbose). I do agree that we should have done the same for the other traits in the file.